### ISSUE TITLE
Adaptive Aggregation Networks for Class-Incremental Learning 

### ISSUE THREAD

**Issue by chaerinkong** [reyllama/MIPAL-CL#8]
## TL; DR

- Two parallel blocks for each residual layer: plasticity (fully learnable) and stability (partially learnable)

## Motivation

- Stability-plasticity dilemma

## Formulation

![image](https://user-images.githubusercontent.com/46308407/158307927-f985a363-83d8-41dc-996b-d1afcd5a39e1.png)
![image](https://user-images.githubusercontent.com/46308407/158307938-85bec110-23ab-4e67-afb3-acd8609a2e87.png)

- Keep two parallel residual blocks for each layer, one for plasticity and the other for stability.
- Plasticity: whole conv block is learnable
- Stability: partially learnable upto weight scaling (for each convolutional kernel)

### BOP (Bilevel Optimization Program)

- Update weighting modulator \alpha and model parameters alternately
- For upper update (\alpha), sample equal size of data from memory and current task
- For lower update (parameters), use all of current task data along with sampled memory

![image](https://user-images.githubusercontent.com/46308407/158308385-20f9e3a8-bbe3-40ee-8882-63c4c9415976.png)

## Evaluation

![image](https://user-images.githubusercontent.com/46308407/158308411-4a1b2acb-64ce-417d-955c-821eafea9441.png)
![image](https://user-images.githubusercontent.com/46308407/158308430-dd223f47-3801-448f-b748-73615ba81f05.png)
![image](https://user-images.githubusercontent.com/46308407/158308451-7c32100f-4004-4830-a5dc-684a57ab8a2e.png)
![image](https://user-images.githubusercontent.com/46308407/158308476-d3df99ae-d6c1-45ec-a078-29d5432f857e.png)
